src/kokkos_abstraction.hpp

```diff
@@ -635,6 +635,36 @@ struct SpaceInstance<Kokkos::Cuda> {
 };
 #endif
 
+struct DeviceDeleter {
+  template <typename T>
+  void operator()(T *ptr) {
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<DevExecSpace>(0, 1),
+        KOKKOS_LAMBDA(const int i) { ptr->~T(); });
+    Kokkos::kokkos_free<DevMemSpace>(ptr);
+  }
+};
+
+template <typename T>
+std::unique_ptr<T, DeviceDeleter> DeviceAllocate() {
```

Comment:
This is cool. I knew this was possible with unique pointers, but I didn't fully understand how to make it work.